[PyTorch] Add test_modules test for TransformerEncoderLayer fast path

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78269
* __->__ #78268

Extend the existing TransformerEncoderLayer test to cover the fast path.

Differential Revision: [D36564009](https://our.internmc.facebook.com/intern/diff/D36564009/)

swolchok committed:
[PyTorch] Add test_modules test for TransformerEncoderLayer fast path

Extend the existing TransformerEncoderLayer test to cover the fast path.

Differential Revision: [D36564009](https://our.internmc.facebook.com/intern/diff/D36564009/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add test_modules test for TransformerEncoderLayer fast path"


Extend the existing TransformerEncoderLayer test to cover the fast path.

Differential Revision: [D36564009](https://our.internmc.facebook.com/intern/diff/D36564009/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add test_modules test for TransformerEncoderLayer fast path"


Extend the existing TransformerEncoderLayer test to cover the fast path.

Differential Revision: [D36564009](https://our.internmc.facebook.com/intern/diff/D36564009/)

[ghstack-poisoned]

jbschlosser commented on torch/testing/_internal/common_modules.py:
```diff
@@ -678,6 +678,29 @@ def module_inputs_torch_nn_TransformerEncoderLayer(module_info, device, dtype, r
                 desc='no_batch_dim'
             ))
 
+    # test fast path. We abuse the fact that we can supply an arbitrary reference_fn to
+    # force the fast path to run -- we cannot call .eval() through the ordinary
+    # ModuleInfo interface, so the "reference" is actually the code under test.
```
We should just fix the `ModuleInfo` interface so it's flexible enough to do what you need here.

Maybe an enum on `ModuleInput` specifying whether it should be used for training / evaluation / both (default) ? If this works okay, I'll open a PR for the support.

jbschlosser replied:
Addressing this in #78735.

jbschlosser replied:
@swolchok FYI this landed. Tests now run in both train and eval, and you can add eval-only sample inputs

swolchok committed:
Update on "[PyTorch] Add test_modules test for TransformerEncoderLayer fast path"


Extend the existing TransformerEncoderLayer test to cover the fast path.

Differential Revision: [D36564009](https://our.internmc.facebook.com/intern/diff/D36564009/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add test_modules test for TransformerEncoderLayer fast path"


Extend the existing TransformerEncoderLayer test to cover the fast path.

Differential Revision: [D36564009](https://our.internmc.facebook.com/intern/diff/D36564009/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add test_modules test for TransformerEncoderLayer fast path"


Extend the existing TransformerEncoderLayer test to cover the fast path.

Differential Revision: [D36564009](https://our.internmc.facebook.com/intern/diff/D36564009/)

[ghstack-poisoned]

zrphercule reviewed (approved):
LGTM, Thanks!

swolchok commented:
@pytorchbot merge
